include/ensmallen_bits/lbfgs/lbfgs_impl.hpp

```diff
@@ -417,34 +440,43 @@ L_BFGS::Optimize(FunctionType& function,
     // least one descent step.
     // TODO: to speed this up, investigate use of arma::norm2est() in Armadillo
     // 12.4
-    if (arma::norm(gradient, 2) < minGradientNorm)
+    //
+    // If we are using FP16, norm() is not currently implemented, so we have to
+    // use norm2est() instead.
+    #if defined(ARMA_HAVE_FP16)
+    const ElemType gradNorm = std::is_same<ElemType, arma::fp16>::value ?
+        norm2est(gradient) : norm(gradient, 2);
```

Comment:
@rcurtin This sets off a red flag for me.  The norm() and norm2est() functions are closely related, but are not the same thing.  They can end up with potentially different results, meaning we have an inconsistency that has nothing to do with the reduced precision of the fp16 element type.

I'd suggest to use norm2est() regardless of the element type, though that would entail bumping the minimum  version of Armadillo.

Alternatively, we can add a workaround directly in Armadillo, where norm(x,2) is implemented for fp16 via internally converting the given matrix to fp32 format.  This in turn allows use of LAPACK functions.  Any modern hardware that does fp16 can also do fp32.

Reply:
Since norm(x, 2) is now implemented in Armadillo, I reverted to `norm()` in 1d963d6c.